### ISSUE TITLE
Schematron support

### ISSUE THREAD

**Issue by tofi86** [xspec/xspec#75]
XSpec currently supports testing XSLT and XQuery.

@AndrewSales had a talk at XML Prague unconference day (in the Schematron Users Meetup) about how to test Schematron schemas and came up with XSpec as one possibility. Althought, there are other tools yet, I like the idea of testing with XSpec.

However, I've no idea yet how to implement it... ;-) This is just a ticket to open discussion...

**Comment by tofi86** [comment]
Link list:
* https://github.com/sutf

**Comment by AndrewSales** [comment]
Thanks, Tobias.
I haven't looked at it myself in any detail, but Tony Graham surfaced this: https://github.com/martian-a/alder.

**Comment by cirulls** [comment]
I like the idea and I can see how this could be useful for Schematron users. Could you post some examples of Schematron schemas and what would you like to test in those schemas? This would help to define the specs for supporting Schematron. 

I can't promise this will be developed but we can definitely look at ways of doing in and, if there is interest, other people may want to contribute or chip in with ideas.

**Comment by tofi86** [comment]
> I haven't looked at it myself in any detail, but Tony Graham surfaced this: https://github.com/martian-a/alder.

@AndrewSales @tgraham-antenna that looks promising... added to the link list!

@cirulls there are some examples in the SUTF repository https://github.com/sutf/samples by @kosek

**Comment by vincentml** [comment]
I created a RelaxNG model that I am using to write unit tests for Schematron, which are then converted via XSLT to XSpec for execution. I would be happy to share what I have.

**Comment by AndrewSales** [comment]
That sounds very interesting - please do share.

**Comment by cirulls** [comment]
Yes, please do share it. Feel free to send a pull request or to post a link to it. It would be useful if you add instructions on how to run it so that we can try it out.

**Comment by vincentml** [comment]
I've started merging my RelaxNG and XSLT for Schematron into a fork of the XSpec repository, and am trying to update xspec.bat so that it can all run with one command. This is taking a little longer than I expected. I'll send a pull request when it's working.

**Comment by vincentml** [comment]
Just an update: I added a few more features for Schematron support than I had originally intended so that is the reason for the delay. There are just a few tests left to write. I will be able to send a pull request soon.

**Comment by cirulls** [comment]
Thanks! We look forward to seeing your pull request! Thanks for writing the tests as well.

**Comment by vincentml** [comment]
I've sent the pull request #105 for adding Schematron support to XSpec. If there are any questions or feedback please let me know.

**Comment by tofi86** [comment]
Wow, awesome @vincentml! 👏 
I'll try to have a look soon...

**Comment by vincentml** [comment]
[Schematron + XSpec.pptx](https://github.com/xspec/xspec/files/961629/Schematron.XSpec.pptx)

I gave a short presentation on Schematron testing with XSpec in the open sessions at JATS-Con yesterday. These are the slides from that presentation.

**Comment by cirulls** [comment]
Thanks for spreading the word and sharing the slides, they will be very useful as a basis for some documentation on the wiki once the new feature is merged into the ```master``` branch.

I have not followed this pull request in detail but I've already heard several people interested in it so I am eager to make it available. Do you  know what it is left to do in order to push it into ```master```?

**Comment by tofi86** [comment]
Thanks for sharing your presentation, @vincentml! And thanks for giving a shout via mail in advance of your talk. Unfortunately the stream didn't work very well in Germany and I wasn't able to follow the conference live...

**Comment by vincentml** [comment]
@cirulls At this point I don’t have anything left to add. I am only making changes in response to feedback. 

In regard to documentation, a file I added to the tutorial folder “tutorial/schematron/schematron.txt” is intended to be a wiki page. This file could be removed from the repository and added to the wiki. The slides could also be used in the documentation.

**Comment by cirulls** [comment]
Thanks @vincentml.

@AirQuick: you reviewed the pull request submitted by @vincentml and suggested some changes. Are you happy with them? Do you see anything else that should be absolutely added/changed before the pull request is merged into ```master```? If I remember well from the pull request thread there are already both documentation and tests for the pull request.

**Comment by AirQuick** [comment]
> Do you see anything else that should be absolutely added/changed before the pull request is merged into master?

No. Since the new modules are built around the existing ones with minimal impact, the pull request looks safe enough to be merged.
My only concern is that as far as I see in the pull request thread, no feedback has been made for the core features. But maybe we can just release it and call for feedback like "New feature… may change!" as @vincentml stated in the presentation.

**Comment by cirulls** [comment]
Thanks to @vincentml for the implementation and to @AirQuick for reviewing it. I'm going to merge it into ```master```. I'm planning to write some tutorial on the wiki in the next days using the documentation provided by @vincentml so that I can familiarize myself with this new feature. I'm planning to mention it at [XML London 2017|http://xmllondon.com/] so that we can spread the word about it and hopefully get feedback from users. Your contributions are also going to be mentioned in the final conference paper.

**Comment by tofi86** [comment]
Awesome! 👏 🎉  Just 3 months from a loose idea to implementation. I'm really looking forward to using and testing this! :-)

**Comment by vincentml** [comment]
The video recording of JATS-Con is now available. The presentation I gave on XSpec Schematron testing is on Day 2 at about 4:43 in. Here is the [link](https://videocast.nih.gov/Summary.asp?File=23267&bhcp=1&start=10350).

**Comment by cirulls** [comment]
Awesome! Very interesting presentation, concise and easy to follow. I added the link in the [wiki](https://github.com/xspec/xspec/wiki/Getting-Started-with-XSpec-and-Schematron#going-further).
